### PR TITLE
look up spec commits concurrently

### PR DESCRIPTION
**Pull request by brendoncarroll**

### CODE REVIEW

File: src/server/pps/server/api_server.go [hunk 1]
```diff
@@ -2492,37 +2492,79 @@ func (a *apiServer) listPipeline(pachClient *client.APIClient, request *pps.List
...
-				pipelineInfo, err = ppsutil.GetPipelineInfo(pachClient, name, ptr)
-				if err != nil {
+				if err := func() error {
+					mu.Lock()
+					defer mu.Unlock()
+					// the filtering shares that buffer thing, and it's CPU bound so why not do it with the lock
+					if !filterPipeline(pinfo) {
+						return nil
+					}
+					return f(pinfo)
```
Comment (Tryneus) [inline src/server/pps/server/api_server.go:2551]: The only problem I see is that we may call `f` after it has already errored.  Perhaps we should have a bool to bail out after acquiring `mu` if an error has occurred (or the `eg` has been canceled)?

**Review by Tryneus — CHANGES_REQUESTED** [review]
LGTM with one concern

File: src/server/pps/server/api_server.go (hunk 1 above)
Comment (brendoncarroll) [reply src/server/pps/server/api_server.go:2551]: Is there a guarantee that f won't be called after it errors?  f won't be called after the function returns.  But I think the caller already doesn't know how many times f will be called.

To get the behavior you're describing, I guess we add another channel and then check the context for cancellation?

**Comment by Tryneus** [comment]
LGTM

**Pull request merged**